C/services/north/north_api.cpp

```diff
@@ -0,0 +1,465 @@
+ *
+ * @param response	The HTTP Response to send
+ * @param request	The HTTP Request
+ */
+static void getDebuggerBufferWrapper(Response response, Request request)
+{
+	if (api)
+		api->getDebuggerBuffer(response, request);
+}
+
+/**
+ * Wrapper for the PUT debugger isolate API call
+ *
+ * @param response	The HTTP Response to send
+ * @param request	The HTTP Request
+ */
+static void isolateDebuggerWrapper(Response response, Request request)
+{
+	if (api)
+		api->isolateDebugger(response, request);
+}
+
+/**
+ * Wrapper for the PUT debugger suspend API call
+ *
+ * @param response	The HTTP Response to send
+ * @param request	The HTTP Request
+ */
+static void suspendDebuggerWrapper(Response response, Request request)
+{
+	if (api)
+		api->suspendDebugger(response, request);
+}
+
+/**
+ * Wrapper for the PUT step debugger API call
+ *
+ * @param response	The HTTP Response to send
+ * @param request	The HTTP Request
+ */
+static void stepDebuggerWrapper(Response response, Request request)
+{
+	if (api)
+		api->stepDebugger(response, request);
+}
+
+/**
+ * Wrapper for the PUT replay debugger API call
+ *
+ * @param response	The HTTP Response to send
+ * @param request	The HTTP Request
+ */
+static void replayDebuggerWrapper(Response response, Request request)
+{
+	if (api)
+		api->replayDebugger(response, request);
+}
+
+/**
+ * Wrapper for the GET state debugger API call
+ *
+ * @param response	The HTTP Response to send
+ * @param request	The HTTP Request
+ */
+static void stateDebuggerWrapper(Response response, Request request)
+{
+	if (api)
+		api->stateDebugger(response, request);
+}
+
+/**
+ * Wrapper for thread creation that is used to start the API
+ */
+static void startService()
+{
+	api->startServer();
+}
+
+/**
+ * North API class constructor
+ *
+ * @param service	The NorthService class this is the API for
+ */
+NorthApi::NorthApi(NorthService *service) : m_service(service), m_thread(NULL)
+{
+	m_logger = Logger::getLogger();
+	m_server = new HttpServer();
+	m_server->config.port = 0;
+	m_server->config.thread_pool_size = 1;
+
+	// Add the debugger entry points
+	m_server->resource[DEBUG_ATTACH]["PUT"] = attachDebuggerWrapper;
+	m_server->resource[DEBUG_DETACH]["PUT"] = detachDebuggerWrapper;
+	m_server->resource[DEBUG_BUFFER]["POST"] = setDebuggerBufferWrapper;
+	m_server->resource[DEBUG_BUFFER]["GET"] = getDebuggerBufferWrapper;
+	m_server->resource[DEBUG_ISOLATE]["PUT"] = isolateDebuggerWrapper;
+	m_server->resource[DEBUG_SUSPEND]["PUT"] = suspendDebuggerWrapper;
+	m_server->resource[DEBUG_STEP]["PUT"] = stepDebuggerWrapper;
+	m_server->resource[DEBUG_REPLAY]["PUT"] = replayDebuggerWrapper;
+	m_server->resource[DEBUG_STATE]["GET"] = stateDebuggerWrapper;
+
+	api = this;
+	m_thread = new thread(startService);
+}
+
+/**
+ * Destroy the API.
+ *
+ * Stop the service and wait fo rthe thread to terminate.
+ */
+NorthApi::~NorthApi()
+{
+	if (m_thread)
+	{
+		m_server->stop();
+		m_thread->join();
```

Comment:
Do you need to call _m_thread->joinable()_ first?

Reply:
No, join will return if the thread has already completed and our threads are not default constructed so they will always be joinable if running.